### PR TITLE
feat(runtime/go): smart ShouldRebuild via go list -deps

### PR DESCRIPTION
**Pull request by alessandrolattao**
## Why

The Go runtime today rebuilds every Lambda in the same `go.mod` tree whenever any file changes. That's fine with the documented layout (one `go.mod` per handler), but as soon as a project chooses a single-module monorepo with many handlers — share a `services/shared` package across N Lambdas, keep one set of dependencies, one `go.mod tidy`, etc. — `sst dev` becomes unusable: editing one handler triggers a rebuild of every other handler in the module.

Side effect of one-go.mod-per-Lambda also pays a recurring cost on every CI lint and CI test pass that scales linearly with the Lambda count, since each handler is its own `golangci-lint` + `go test` invocation.

This is the same problem the Node runtime already solved: it uses esbuild's metafile to know exactly which files compile into each bundle and rebuilds only the Lambdas whose actual deps changed.

## What

Capture each handler's transitive import graph at `Build` time via `go list -deps -json` and use it as the rebuild scope. `ShouldRebuild` fires only when the changed file is in that handler's graph. Works in every layout (per-Lambda, per-domain, single-module monorepo).

Side fix: `go build` arguments now use a `./` prefix on the source path so layouts where the source is a sub-path don't fail with `package <path> is not in std`. With the legacy one-go.mod-per-Lambda layout `src` was always `.`, so the bug never surfaced.

The previous directory-scope behavior was deliberately removed rather than kept as a fallback. In a single-module repo it would silently fan out to every Lambda when `captureDeps` failed, hiding a real problem behind a worse one. A `captureDeps` failure now logs a warning with full context (`goflags`, `gomodcache`) and disables `ShouldRebuild` for that function until the next successful `Build`.

## Tests

\`pkg/runtime/golang/golang_test.go\` — pure unit tests for \`Properties\` JSON contract, the goarch mapping, \`ShouldRebuild\` edge cases (non-go file, missing graph, file hit/miss, relative path resolution), the \`GOMODCACHE\`/stdlib filter via a fixture JSON stream, a 50×50 race-detector concurrency probe, and a cross-OS \`isUnderDir\` helper.

\`pkg/runtime/golang/build_test.go\` — hermetic integration tests against the real \`go list\` toolchain (env scrubbed: \`GOPROXY=off\`, \`GOTOOLCHAIN=local\`, redirected \`GOMODCACHE\`/\`GOCACHE\`/\`HOME\`). Skipped under \`-short\` and when \`go\` is not in \`PATH\`.

\`go test ./pkg/runtime/golang/... -race\`: 1.6s full, 1.0s under \`-short\`. \`go vet\` clean.

## Tested manually

Validated end-to-end on a real repo with ~60 Go Lambdas consolidated into one root \`services/go.mod\`. With this patch:

- editing a single Lambda's \`handler.go\` rebuilds only that Lambda;
- editing a file in the shared package rebuilds only the Lambdas that import it (cross-Lambda fan-out is correct);
- editing a \`_test.go\` triggers no rebuild (test files don't compile into the binary);
- \`bun run go-lint\` time on the same repo dropped from ~10s to <1s warm cache (one \`golangci-lint\` invocation over the whole module instead of 60).

### CODE REVIEW

File: pkg/runtime/golang/golang.go [hunk 1]
```diff
@@ -130,18 +180,102 @@ func (r *Runtime) Run(ctx context.Context, input *runtime.RunInput) (runtime.Wor
...
 	}
-	slog.Info("checking if file needs to be rebuilt", "file", file, "match", match)
-	rel, err := filepath.Rel(match, file)
+	abs, err := filepath.Abs(file)
+	if err != nil {
+		return false
+	}
+	if _, found := files[abs]; !found {
+		return false
+	}
```
[the remark this answers is not in the document]
Comment (alessandrolattao) [reply pkg/runtime/golang/golang.go:215]: Done — added per-handler `pkgDirs` populated in `parseGoListOutput` from `go list`'s package dirs (`golang.go:26, 280`), and a third branch in `ShouldRebuild` that fires when a changed `.go` file is under one of those dirs (`golang.go:221-228`). The new file enters the captured set on the next Build. Covered by `TestShouldRebuild_NewFileInTrackedPackage`.

File: pkg/runtime/golang/golang.go [hunk 2]
```diff
@@ -103,7 +129,30 @@ func (r *Runtime) Build(ctx context.Context, input *runtime.BuildInput) (*runtim
...
+	r.directories[input.FunctionID] = absRoot
+	if depErr != nil {
+		delete(r.files, input.FunctionID)
+	} else {
+		r.files[input.FunctionID] = deps
+	}
+	r.mut.Unlock()
+
+	if depErr != nil {
+		slog.Warn("failed to capture go deps; ShouldRebuild disabled until next build",
```
[the remark this answers is not in the document]
Comment (alessandrolattao) [reply pkg/runtime/golang/golang.go]: Done — capture failure now keeps the previous graph (`golang.go:146-150`): `r.files` / `r.pkgDirs` are only assigned when `depErr == nil`. On failure we `slog.Warn` with full context (`fn`, `err`, `root`, `src`, `goflags`, `gomodcache`) and `ShouldRebuild` keeps using the last good graph until the next successful Build.

File: pkg/runtime/golang/golang.go [hunk 3]
```diff
@@ -130,18 +180,102 @@ func (r *Runtime) Run(ctx context.Context, input *runtime.RunInput) (runtime.Wor
...
+
+func (r *Runtime) captureDeps(ctx context.Context, root, src string, env []string) (map[string]struct{}, error) {
+	cmd := exec.CommandContext(ctx, "go", "list", "-deps", "-json", "./"+src)
+	cmd.Dir = root
+	cmd.Env = env
+	out, err := cmd.Output()
+	if err != nil {
+		return nil, err
+	}
+	return parseGoListOutput(strings.NewReader(string(out)), r.resolveGoModCache(env))
```
[the remark this answers is not in the document]
Comment (alessandrolattao) [reply pkg/runtime/golang/golang.go:251]: Skipped — kept `cmd.Output()` for now. `go list -deps -json` for a single handler stays well under 1 MB even on large monorepos (validated end-to-end on a ~60-Lambda repo), and the buffered path collapses the error surface from four (pipe / start / parse / wait) into one. Happy to revisit if a real profile shows it matters.

File: pkg/runtime/golang/golang_test.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			if r.ShouldRebuild("fn", tt.file) {
+				t.Errorf("expected false for non-.go file %q", tt.file)
+			}
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (alessandrolattao) [reply pkg/runtime/golang/golang_test.go]: Done — the test was renamed to `TestShouldRebuild_UntrackedFiles` and now asserts only that genuinely-untracked files don't rebuild. Added explicit coverage for the cases this comment flagged: `TestShouldRebuild_EmbedAssetInGraph` (a tracked embed asset triggers rebuild) and `TestShouldRebuild_GoModAndGoSum` (handler's `go.mod` and `go.sum` trigger rebuild).

File: pkg/runtime/golang/golang.go [hunk 5]
```diff
@@ -130,18 +180,102 @@ func (r *Runtime) Run(ctx context.Context, input *runtime.RunInput) (runtime.Wor
 	}, nil
 }
 
+// ShouldRebuild reports whether a file change must trigger a rebuild
+// of the given Lambda. Returns true only when the file is part of the
+// handler's transitive import graph captured by the last Build.
 func (r *Runtime) ShouldRebuild(functionID string, file string) bool {
 	if !strings.HasSuffix(file, ".go") {
```
[the remark this answers is not in the document]
Comment (alessandrolattao) [reply pkg/runtime/golang/golang.go]: Done — the `.go` suffix check now only guards the third branch (new file in tracked pkg dir, `golang.go:221`), where it's needed because that branch matches by directory containment and would otherwise accept editor swap files like `4913` that live in a tracked dir. The two prior branches (file in captured graph, handler's `go.mod` / `go.sum`) match by exact path with no suffix filter, so embed assets and module files trigger rebuild as expected. Covered by `TestShouldRebuild_EmbedAssetInGraph` and `TestShouldRebuild_GoModAndGoSum`.

**Comment by alessandrolattao** [comment]
I've just realized that the current version fails to trigger a rebuild for Lambdas when code shared via workspaces/replaces is modified, even if the Lambda depends on it. This update also fixes that issue.
